TP 2 — Préparation des données avec Python
Exercice 1 — Exploration et diagnostic
Affichez les dimensions du dataset
Affichez les types de données avec info()
Comptez les valeurs manquantes par colonne (nombre et pourcentage)
Comptez le nombre de doublons
Affichez les valeurs uniques de Region — repérez les incohérences
Affichez les statistiques de Total Deaths — repérez les anomalies

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv("catnat_dirty.csv")

In [3]:
#Affichez les dimensions du dataset
df.shape

(17510, 18)

In [4]:
# Affichez les types de données avec info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17510 entries, 0 to 17509
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   DisNo.                   17510 non-null  object 
 1   Country                  17510 non-null  object 
 2   Region                   17161 non-null  object 
 3   Subregion                17510 non-null  object 
 4   Disaster Type            17510 non-null  object 
 5   Disaster Subtype         17510 non-null  object 
 6   Disaster Subgroup        17510 non-null  object 
 7   Event Name               3975 non-null   object 
 8   Start Year               17510 non-null  object 
 9   Start Month              16599 non-null  float64
 10  Total Deaths             12593 non-null  float64
 11  No. Injured              4507 non-null   float64
 12  Total Affected           12951 non-null  float64
 13  No. Homeless             2521 non-null   float64
 14  Total Damage ('000 US$

In [5]:
#Comptez le nombre de doublons
df.duplicated().sum()

np.int64(150)

In [6]:
# Comptez les valeurs manquantes par colonne (nombre et pourcentage)
df.isnull().sum()

DisNo.                         0
Country                        0
Region                       349
Subregion                      0
Disaster Type                  0
Disaster Subtype               0
Disaster Subgroup              0
Event Name                 13535
Start Year                     0
Start Month                  911
Total Deaths                4917
No. Injured                13003
Total Affected              4559
No. Homeless               14989
Total Damage ('000 US$)    11880
Magnitude                  12261
Latitude                   14689
Longitude                  14689
dtype: int64

In [7]:
# Affichez les valeurs uniques de Region — repérez les incohérences
df['Region'].value_counts()

Region
Asia          3754
Americas      2359
Africa        1688
Europe        1201
ASIA           792
asia           763
americas       504
AMERICAS       486
Asia           429
 Asia          396
Oceania        396
AFRICA         375
africa         370
 Asia          282
EUROPE         265
 Americas      248
europe         232
Americas       231
 Africa        202
Africa         181
 Americas      178
 Africa        140
 ASIA          116
Europe         105
 Europe        100
OCEANIA         84
 asia           82
ASIA            77
oceania         75
 Europe         75
asia            69
 asia           61
 ASIA           60
 AMERICAS       57
americas        56
AMERICAS        46
 americas       44
 AMERICAS       42
AFRICA          42
Oceania         41
 AFRICA         38
 Oceania        38
 africa         37
africa          37
 AFRICA         34
 americas       32
 europe         31
EUROPE          31
europe          30
 EUROPE         26
 africa         25
 Oceania        22
 euro

In [8]:
#Affichez les statistiques de Total Deaths — repérez les anomalies
df['Total Deaths'].describe()

count    1.259300e+04
mean     2.751663e+03
std      6.577989e+04
min     -3.600000e+04
25%      5.000000e+00
50%      1.800000e+01
75%      6.100000e+01
max      3.700000e+06
Name: Total Deaths, dtype: float64

Questions
Combien y a-t-il de doublons ? 150
Quelles colonnes ont le plus de valeurs manquantes ? Latitude et Longitude avec 14689
Combien de variantes différentes pour "Asia" ? 12
Y a-t-il des valeurs négatives dans Total Deaths ? Oui on peut voir 'min -3.600000e+04"


Exercice 2 — Suppression des doublons

In [9]:
# Identifier les doublons
print(df[df.duplicated()].head())  # Affiche les doublons, sans les originaux
df[df.duplicated()]

print(df[df.duplicated(keep=False)].head())  # Inclut les originaux

# Supprimer les doublons
df = df.drop_duplicates()
print("Nombre de doublons restants :", df.duplicated().sum())


             DisNo.     Country  Region           Subregion  \
1210  2021-0596-PAK    Pakistan    Asia       Southern Asia   
1959  2019-0655-BGD  Bangladesh   Asia        Southern Asia   
2157  2011-9629-LSO    Lesotho   africa  Sub-Saharan Africa   
3420  2004-0668-GAB       Gabon  Africa  Sub-Saharan Africa   
3494  2002-0406-JPN       JAPAN    Asia        Eastern Asia   

            Disaster Type Disaster Subtype Disaster Subgroup     Event Name  \
1210                Flood  Flood (General)      Hydrological            NaN   
1959  Extreme temperature        Cold wave    Meteorological            NaN   
2157              DROUGHT          Drought    Climatological            NaN   
3420             Epidemic    Viral disease        Biological  Typhoid fever   
3494                Storm  Storm (General)    Meteorological            NaN   

     Start Year  Start Month  Total Deaths  No. Injured  Total Affected  \
1210       2021          9.0       19000.0          4.0             4.0

Exercice 3

In [10]:
# Vérifiez le type de Start Year
print(df['Start Year'].dtypes)
# Affichez quelques valeurs problématiques (contenant "Year" ou "AD")
valpb = df["Start Year"].astype(str).str.contains("Year|AD", na=False)

print(df.loc[valpb, "Start Year"].head(10))
# Nettoyez la colonne : supprimez le texte et convertissez en numérique
df["Start Year"] = df["Start Year"].str.replace("Year", "", regex=False)
df["Start Year"] = df["Start Year"].str.replace("AD", "", regex=False)
df['Start Year'] = pd.to_numeric(df['Start Year'], errors='coerce')

# Vérifiez le résultat
print(df['Start Year'].dtypes)

object
26        1982 AD
138     Year 2012
480       2021 AD
506     Year 2017
634     Year 2003
708     Year 1964
924       2000 AD
947       2017 AD
1013      2019 AD
1075    Year 1999
Name: Start Year, dtype: object
int64


Exercice 4 

In [11]:
# Pour Region (peu de nulls) : supprimez les lignes avec null
df.dropna(subset=['Region'])          
# Pour Event Name : remplacez les nulls par "Non nommé"
df['Event Name'].fillna('Non nommé')
# Pour Total Deaths : décidez d'une stratégie et appliquez-la
df['Total Deaths'].fillna('Non dit')
# Pour Magnitude : laissez les nulls (n'a pas de sens pour tous les types)
# Vérifiez le résultat
print(df['Region'].isnull().sum())
print(df['Event Name'].isnull().sum())
print(df['Total Deaths'].isnull().sum())
print(df['Magnitude'].isnull().sum())

346
13415
4878
12153


Exercice 5 — Traitement des outliers

In [12]:
# Identifiez les valeurs négatives dans Total Deaths
df[df['Total Deaths'] < 0]
# Identifiez les valeurs aberrantes dans Magnitude (> 10)
df[df['Magnitude'] > 10]
# Décidez : supprimer ou corriger ?
# Appliquez le traitement
df = df[df['Magnitude'] >= 0]
# Vérifiez le résultat
print(df[df['Magnitude'] > 10])

              DisNo.                           Country    Region  \
5      1997-0566-BOL  Bolivia (Plurinational State of)  Americas   
32     1996-0215-LAO  Lao People's Democratic Republic      Asia   
34     2002-0379-USA          United States of America  AMERICAS   
40     1995-0198-THA                          Thailand     Asia    
42     1996-0298-IND                             India      Asia   
...              ...                               ...       ...   
17479  2008-0043-MDG                        MADAGASCAR    Africa   
17490  2003-0447-SOM                           Somalia    Africa   
17499  2022-0099-MDG                        Madagascar    Africa   
17501  2018-0123-AUS                         Australia   Oceania   
17502  1991-0131-MDV                          Maldives      Asia   

                             Subregion Disaster Type  Disaster Subtype  \
5      Latin America and the Caribbean         flood    Riverine flood   
32                  South-eastern A

Exercice 6

In [13]:
df['Region'] = df['Region'].str.strip().str.lower()
df['Region'].value_counts()
df['Disaster Type']= df["Disaster Type"].str.strip().str.lower()
df['Disaster Type'].value_counts()

df['Country']= df['Country'].str.title().str.strip()
df['Country'].replace({'USA': 'United States Of America'})
df['Country'].value_counts()


Country
China                                    358
United States Of America                 270
Indonesia                                209
India                                    201
Philippines                              177
                                        ... 
Côte  D’Ivoire                             1
Tuvalu                                     1
Saint  Vincent  And  The  Grenadines       1
Venezuela  (Bolivarian  Republic  Of)      1
Gabon                                      1
Name: count, Length: 246, dtype: int64

Exercice 7

In [14]:
# Créez une colonne Decennie à partir de Start Year
df["Decennie"] = (df["Start Year"] // 10) * 10

# Créez une colonne Has_Deaths (booléen : True si Total Deaths > 0)
df["Has_Deaths"] = df["Total Deaths"] > 0

# Vérifiez vos nouvelles colonnes
print(df[["Start Year", "Decennie", "Total Deaths", "Has_Deaths"]].head(10))
print(df.dtypes)
print(df["Has_Deaths"].value_counts())

    Start Year  Decennie  Total Deaths  Has_Deaths
4         2008      2000           1.0        True
5         1997      1990          16.0        True
19        1977      1970         589.0        True
22        1922      1920        1000.0        True
26        1982      1980           3.0        True
32        1996      1990          30.0        True
34        2002      2000           NaN       False
38        2005      2000           5.0        True
40        1995      1990         231.0        True
42        1996      1990           6.0        True
DisNo.                      object
Country                     object
Region                      object
Subregion                   object
Disaster Type               object
Disaster Subtype            object
Disaster Subgroup           object
Event Name                  object
Start Year                   int64
Start Month                float64
Total Deaths               float64
No. Injured                float64
Total Affected     

Exercice 8

In [21]:
# Sélectionnez les colonnes utiles pour l'analyse
df_export = df[["DisNo.","Event Name", "Start Year", "Region","Subregion", "Country", "Disaster Type", "Total Deaths", "Magnitude", "Decennie", "Has_Deaths"]]
# Renommez les colonnes si nécessaire (noms clairs, sans espaces)
df_export = df_export.rename(columns={'Event Name' : 'Event_Name','Start Year': 'Start_Year','Disaster Type' : 'Disaster_Type', 'Total Deaths': 'Total_Deaths'})
# Faites une vérification finale avec info() et head()
print(df_export.info())
print(df_export.head())
# Exportez en CSV : catnat_clean.csv
df_export.to_csv("catnat_clean.csv", index=False)


<class 'pandas.core.frame.DataFrame'>
Index: 5060 entries, 4 to 17507
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DisNo.         5060 non-null   object 
 1   Event_Name     902 non-null    object 
 2   Start_Year     5060 non-null   int64  
 3   Region         4942 non-null   object 
 4   Subregion      5060 non-null   object 
 5   Country        5060 non-null   object 
 6   Disaster_Type  5060 non-null   object 
 7   Total_Deaths   3972 non-null   float64
 8   Magnitude      5060 non-null   float64
 9   Decennie       5060 non-null   int64  
 10  Has_Deaths     5060 non-null   bool   
dtypes: bool(1), float64(2), int64(2), object(6)
memory usage: 439.8+ KB
None
           DisNo. Event_Name  Start_Year    Region  \
4   2008-0275-JPN        NaN        2008      asia   
5   1997-0566-BOL        NaN        1997  americas   
19  1977-0142-IRN        NaN        1977      asia   
22  1922-0001-CHL        NaN   